### DISCUSSION TITLE
How to create dynamic matrix value?

### DISCUSSION THREAD

**Discussion by oonid** [community/community#27150] [Actions]
<p>hello,</p>
<p>I know from <a href="https://github.community/t/secrets-in-matrix/17307">this post</a> that the matrix can not be a value from secrets.</p>
<p>any thoughts on how to make matrix value can be dynamic, so I don’t need to change the workflow file every time I add a new app?</p>
<p>so here’s my current <a href="https://github.com/oonid/jagodjango/blob/master/.github/workflows/djangoapps.yml" rel="nofollow noopener">yml file</a>:</p>
<pre><code>jobs:
  build:

    runs-on: ubuntu-latest
    strategy:
      max-parallel: 4
      matrix:
        python-version: [3.6, 3.7, 3.8]
        django-version: [2.2, 3.0.0]
        django-app: [halo, splitviewsmodels]
</code></pre>
<p>actually the django-app list is a directory names inside my project, so I can easily get the names of it, if I can’t set the value from secrets(?)</p>
<p>thank you for your help.</p>

**Comment by BrightRan** [comment]
<p>Currently, we have no any way to make the values of a matrix parameter can be dynamic.</p>
<p>If you want to add new values into the matrix parameter, you have to change the workflow definition every time. You can use the keyword exclude in the matrix job to let job not run for some combinations of the configurations.</p>
<p>There is a ticket reported the similar question as reference: <a href="https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.community%2Ft5%2FGitHub-Actions%2FDynamically-Select-Secrets%2Ftd-p%2F42612&amp;data=02%7C01%7Cv-brira%40microsoft.com%7Cf3a440ac04f44400b25508d794c99fbb%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637141466006933295&amp;sdata=fkYn3txezes9XzEdFIip4oS49gtu1ldi%2FTeQvBVXNmo%3D&amp;reserved=0" rel="nofollow noopener">https://github.community/t5/GitHub-Actions/Dynamically-Select-Secrets/td-p/42612</a></p>

**Comment by maraisr** [comment]
<p>Im actually really keen on this idea.</p>
<p>Im using the matrix to build apps in my mono repo, but ideally I’d only like to build apps that have changed.</p>
<p>So im looking for something like: (beware sudo code)</p>
<pre><code>jobs:
  collect_apps:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v1
      - run: echo "find changed paths"
      - run: |
          # for each app
            echo "::add-matrix name=app::%changed_path%"
          # end for each
  build:
    runs-on: ubuntu-latest
    strategy:
      matrix:
        app: ${{ steps.collect_apps.matrix }}
    env:
      TZ: Australia/Brisbane
    steps:
      - uses: actions/checkout@v1
      - run: yarn install
      - name: Build (${{ matrix.app }})
        id: app_built
        run: |
          yarn workspace ${{ matrix.app }} build;
          echo "::set-output name=dist_dir::apps/${{ matrix.app }}/dist"
      - uses: actions/upload-artifact@v1.0.0
        with:
          name: ${{ matrix.app }}
          path: ${{ steps.app_built.outputs.dist_dir }}
</code></pre>

**Comment by pozil** [comment, marked as answer]
<p>You can now build a dynamic matrix with <code>fromJson()</code><br>
</p><aside class="onebox whitelistedgeneric">
  <header class="source">
      <img src="https://github.blog/wp-content/uploads/2019/01/cropped-github-favicon-512.png?fit=32%2C32" class="site-icon" width="32" height="32">
      <a href="https://github.blog/changelog/2020-04-15-github-actions-new-workflow-features/" target="_blank" rel="nofollow noopener">The GitHub Blog</a>
  </header>
  <article class="onebox-body">
    <div class="aspect-image" style="--aspect-ratio:690/362;"><img src="https://github.blog/wp-content/uploads/2019/03/blog-card.png?fit=5001%2C2626" class="thumbnail" width="690" height="362"></div>

<h3><a href="https://github.blog/changelog/2020-04-15-github-actions-new-workflow-features/" target="_blank" rel="nofollow noopener">GitHub Actions: New workflow features The GitHub Blog</a></h3>

<p>GitHub Actions: New workflow features</p>


  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>

**Comment by oonid** [comment]
<p>thank you!<br>
will check the latest workflow feature.</p>

**Comment by frezbo** [comment]
<p>I got dynamic matrix working by passing in a json object. An example here: <a href="https://github.com/frezbo/dockerfiles/blob/6701b8cffc3b3628ad72b541d2c36ac87e1e9b8f/.github/workflows/ci.yml" class="inline-onebox" rel="noopener nofollow ugc">dockerfiles/ci.yml at 6701b8cffc3b3628ad72b541d2c36ac87e1e9b8f · frezbo/dockerfiles · GitHub</a></p>
